// node-math/index.js
// Exponentials
module.exports.pow = Math.pow;
module.exports.square = function(n) { return n * n };
module.exports.cube = function(n) { return n * n * n };
module.exports.sqrt = Math.sqrt;

// Constants
module.exports.pi = Math.PI;
module.exports.e = Math.E;
module.exports.round = Math.round;
module.exports.ceil = Math.ceil;
module.exports.floor = Math.floor;

// Trignometry
module.exports.cos = Math.cos;
module.exports.tan = Math.tan;
module.exports.sin = Math.sin;

// Checks
module.exports.isEven = function(n) {
    return n % 2 === 0;
}

module.exports.isOdd = function(n) {
    return n % 2 != 0;
}

module.exports.isBin = function(str) {
    return !isNaN(parseInt(str, 2)) && isFinite(str);
}

module.exports.isInt = function(n) {
    return n % 1 === 0;
}

module.exports.isNumeric = function(n) {
    return !isNaN(parseInt(n, 10)) && isFinite(n);
}

// From
module.exports.bin = function(str) {
    return parseInt(str, 2);
}

module.exports.str = function(str) {
    return parseFloat(str);
}

// Mass Number 
module.exports.sum = function() {
    return Array.prototype.slice.call(arguments).reduce(function(pre, curr) {
	return pre + curr;
    });
}

module.exports.product = function() {
    return Array.prototype.slice.call(arguments).reduce(function(pre, curr) {
	return pre * curr;
    });
}

module.exports.min = Math.min;
module.exports.max = Math.max;

// Number Ranges

module.exports.range = function(start, stop, step) {
    if(stop === undefined) {
	stop = start;
	start = 1;
    }
    if(step === undefined) step = 1;
    var x = start || 1, res = [];
    while(x <= stop) {
	res.push(x);
	x += step;
    }
    return res;
}

module.exports.sort = function() {
    return Array.prototype.slice.call(arguments).sort(function(a, b) {
	return a - b;
    });
}

module.exports.median = function() {
    var values = this.sort.apply(this, Array.prototype.slice.call(arguments));
    var half = Math.floor(values.length / 2);
    if(this.isOdd(values.length)) return values[half];
    else return (values[half -1] + values[half]) / 2;
}

module.exports.random(min, max) {
    if(max === undefined) {
	max = min;
	min = 1;
    }
    if(min === undefined && max === undefined) return Math.random();
    return Math.floor(Math.random * (max - min) + min);
}
